rusk-wallet/src/lib/clients.rs

```diff
@@ -0,0 +1,270 @@
+use dusk_jubjub::{BlsScalar, JubJubAffine, JubJubScalar};
+use dusk_pki::ViewKey;
+use dusk_plonk::prelude::Proof;
+use dusk_poseidon::tree::PoseidonBranch;
+use dusk_wallet_core::{UnprovenTransaction, POSEIDON_TREE_DEPTH};
+use phoenix_core::{Crossover, Fee, Note};
+
+use std::sync::Mutex;
+use tokio::runtime::Handle;
+use tokio::task::block_in_place;
+use tonic::transport::Channel;
+
+use crate::errors::CliError;
+
+use crate::rusk_proto::network_client::NetworkClient;
+use crate::rusk_proto::prover_client::ProverClient;
+use crate::rusk_proto::state_client::StateClient;
+
+use crate::rusk_proto::PropagateMessage;
+use crate::rusk_proto::{
+    ExecuteProverRequest, StctProverRequest, WfctProverRequest,
+};
+use crate::rusk_proto::{
+    GetAnchorRequest, GetNotesOwnedByRequest, GetOpeningRequest,
+    GetStakeRequest,
+};
+
+/// Implementation of the ProverClient trait from wallet-core
+#[derive(Debug)]
+pub struct Prover {
+    client: Mutex<ProverClient<Channel>>,
+    network: Mutex<NetworkClient<Channel>>,
+}
+
+impl Prover {
+    pub fn new(
+        client: ProverClient<Channel>,
+        network: NetworkClient<Channel>,
+    ) -> Self {
+        Prover {
+            client: Mutex::new(client),
+            network: Mutex::new(network),
+        }
+    }
+}
+
+impl dusk_wallet_core::ProverClient for Prover {
+    /// Error returned by the prover client.
+    type Error = CliError;
+
+    /// Requests that a node prove the given transaction and later propagates it
+    fn compute_proof_and_propagate(
+        &self,
+        utx: &UnprovenTransaction,
+    ) -> Result<(), Self::Error> {
+        let utx_bytes = utx.to_var_bytes();
+        let msg = ExecuteProverRequest { utx: utx_bytes };
+        let req = tonic::Request::new(msg);
+
+        let mut prover = self.client.lock().unwrap();
+        let tx_bytes = block_in_place(move || {
+            Handle::current()
+                .block_on(async move { prover.prove_execute(req).await })
+        })?
+        .into_inner()
+        .tx;
+
+        // todo: encode message
+        let msg = PropagateMessage { message: tx_bytes };
+        let req = tonic::Request::new(msg);
+
+        let mut net = self.network.lock().unwrap();
+        let _ = block_in_place(move || {
+            Handle::current().block_on(async move { net.propagate(req).await })
+        })?;
+
+        Ok(())
+    }
+
+    /// Requests an STCT proof.
+    fn request_stct_proof(
+        &self,
+        fee: &Fee,
+        crossover: &Crossover,
+        value: u64,
+        blinder: JubJubScalar,
+        address: BlsScalar,
+        signature: Signature,
+    ) -> Result<Proof, Self::Error> {
+        let mut buf = BytesMut::new();
+        buf.put_slice(&fee.to_bytes());
+        buf.put_slice(&crossover.to_bytes());
+        buf.put_slice(&value.to_bytes());
+        buf.put_slice(&blinder.to_bytes());
+        buf.put_slice(&address.to_bytes());
+        buf.put_slice(&signature.to_bytes());
```

Comment:
Could you use the [`dusk_bytes::Write`](https://docs.rs/dusk-bytes/0.1.5/dusk_bytes/trait.Write.html) API?